src/ctapipe/reco/telecope_event_handling.py

```diff
@@ -0,0 +1,145 @@
+"""Helper functions for array-event-wise aggregation of telescope events."""
+
+import numpy as np
+from numba import njit, uint64
+
+__all__ = ["get_subarray_index", "weighted_mean_std_ufunc"]
+
+
+@njit
+def _get_subarray_index(obs_ids, event_ids):
+    n_tel_events = len(obs_ids)
+    idx = np.zeros(n_tel_events, dtype=uint64)
+    current_idx = 0
+    subarray_obs_index = []
+    subarray_event_index = []
+    multiplicities = []
+    multiplicity = 0
+
+    if n_tel_events > 0:
+        subarray_obs_index.append(obs_ids[0])
+        subarray_event_index.append(event_ids[0])
+        multiplicity += 1
+
+    for i in range(1, n_tel_events):
+        if obs_ids[i] != obs_ids[i - 1] or event_ids[i] != event_ids[i - 1]:
+            # append to subarray events
+            multiplicities.append(multiplicity)
+            subarray_obs_index.append(obs_ids[i])
+            subarray_event_index.append(event_ids[i])
+            # reset state
+            current_idx += 1
+            multiplicity = 0
+
+        multiplicity += 1
+        idx[i] = current_idx
+
+    # Append multiplicity of the last subarray event
+    if n_tel_events > 0:
+        multiplicities.append(multiplicity)
+
+    return (
+        np.asarray(subarray_obs_index),
+        np.asarray(subarray_event_index),
+        np.asarray(multiplicities),
+        idx,
+    )
+
+
+def get_subarray_index(tel_table):
+    """
+    Get the obs_ids and event_ids of all subarray events contained
+    in a table of telescope events, their multiplicity and an array
+    giving the index of the subarray event for each telescope event.
+    This requires the telescope events to be SORTED by their corresponding
+    subarray events (meaning by ``["obs_id", "event_id"]``).
+
+    Parameters
+    ----------
+    tel_table: astropy.table.Table
+        table with telescope events as rows
+
+    Returns
+    -------
+    Tuple(np.ndarray, np.ndarray, np.ndarray, np.ndarray)
+        obs_ids of subarray events, event_ids of subarray events,
+        multiplicity of subarray events, index of the subarray event
+        for each telescope event
+    """
+    obs_idx = tel_table["obs_id"]
+    event_idx = tel_table["event_id"]
+    return _get_subarray_index(obs_idx, event_idx)
+
+
+def _grouped_add(tel_data, n_array_events, indices):
+    """
+    Calculate the group-wise sum for each array event over the
+    corresponding telescope events. ``indices`` is an array
+    that gives the index of the subarray event for each telescope event.
+    """
+    combined_values = np.zeros(n_array_events)
+    np.add.at(combined_values, indices, tel_data)
+    return combined_values
+
+
+def weighted_mean_std_ufunc(
+    tel_values,
+    valid_tel,
+    n_array_events,
+    indices,
+    multiplicity,
+    weights=np.array([1]),
```

Comment:
Using 1 is the easiest code, but we should be able to gain performance and the difference in code should not be so large to special case the "no weights" case, right?

Reply:
I am struggling to get the multiplicity including only the valid telescope events (as in what I need to divide `sum_prediction` by) without calculating something like a `sum_of_weights`. Obviously, it would be possible to get them in `StereoMeanCombiner` via `np.count_nonzero(events[<prefix-of-energy-reco>_telescopes])` and parse them into `weighted_mean_std_ufunc`, but I think that would be really ugly and confusing, since this would only be needed for `weights='none'` and I would prefer any other solution. 
I will have a look again on Monday. I'm probably missing something really obvious...